drivers: sensor: bmi08x: fix temperature interface and trigger setting

ttwards opened the pull request:
1. Temperature Interface
According to BMI08x datasheet, temperature reading
requires both MSB and LSB bytes to be read and
processed correctly.

Temp data processing should follow the formula:
Temp in °C = (temp_msb * 8) + (temp_lsb / 32)

This patch implements the correct reading
sequence and calculation method as specified
in the datasheet.

2. Trigger Setting
Previously we set handler and then trigger struct.
However under some situation, as long as we set
the handler, we get into ISR immediately and never
set trigger struct.
I simply changed the sequence.

Testing:
- Verified temperature readings match datasheet
- Tested on stm32f407igh board with BMI08x sensor

Datasheet:
<img width="784" alt="截屏2024-12-01 16 21 40" src="https://github.com/user-attachments/assets/c36af272-3278-49dc-bcca-40bde72f562b">
<img width="765" alt="截屏2024-12-01 16 21 59" src="https://github.com/user-attachments/assets/40701aea-71ec-41c5-8340-75d4f196a939">

ttwards commented:
Code Lints and commit lints are fixed, please check again

XenuIsWatching commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -536,7 +548,7 @@ static int bmi08x_accel_pm_action(const struct device *dev, enum pm_device_actio
 #endif /* CONFIG_PM_DEVICE */
 
-static DEVICE_API(sensor, bmi08x_api) = {
+static const struct sensor_driver_api bmi08x_api = {
```
you must revert this line

ttwards replied:
Commit a553af7 used DEVICE_API to replace defining the api struct

XenuIsWatching commented:
remove the merge commit, you must always rebase

ttwards commented:
> remove the merge commit, you must always rebase, and this should all only be 1 commit

I've removed, thank you.

XenuIsWatching commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,19 +454,30 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
-	temp_micro = BMI08X_TEMP_OFFSET * 1000000ULL + temp_raw * 31250ULL;
-
+	ret = bmi08x_accel_word_read(dev, BMI08X_REG_TEMP_LSB, &temp_raw);
+	temp_int11 |= (temp_raw & 0xE0) >> 5;
+	if (ret < 0) {
```
wouldn't it make sense to check the `ret` first

XenuIsWatching reviewed (changes requested):
please fix the compliance issues

ttwards committed:
drivers: sensor: bmi08x: fix interfaceand trigger

1. Temperature Interface
According to BMI08x datasheet, temperature reading
requires both MSB and LSB bytes to be read and
processed correctly.

Temp data processing should follow the formula:
Temp in °C = (temp_msb * 8) + (temp_lsb / 32)

This patch implements the correct reading
sequence and calculation method as specified
in the datasheet.

2. Trigger Setting
Previously we set handler and then trigger struct.
However under some situation, as long as we set
the handler, we get into ISR immediately and never
set trigger struct.
I simply changed the sequence.

Testing:
- Verified temperature readings match datasheet
- Tested on stm32f407igh board with BMI08x sensor

Fixes: #82375

Signed-off-by: Wenxi Xu <xuwenxi0517@gmail.com>

ttwards commented:
Hmmmm, can anybody please tell me why there are 4 pending checks still...?

kartben commented:
> Hmmmm, can anybody please tell me why there are 4 pending checks still...?

A maintainer needs to kick the CI manually for contributions coming from new contributors. I just did :) thanks!

ttwards commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,19 +454,31 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
+	}
 
+	ret |= bmi08x_accel_word_read(dev, BMI08X_REG_TEMP_LSB, &temp_raw);
+	if(!ret) {
```
solved, thanks! Sorry for not being familiar with pr and code lints....

MaureenHelm commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,19 +454,31 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
+		return -ENODATA;
+	}
 
+	ret |= bmi08x_accel_word_read(dev, BMI08X_REG_TEMP_LSB, &temp_raw);
```
You might not have a valid errno after the bitwise OR operation

ttwards replied:
Corrected

ttwards commented:
Problem proceeding return value is fixed, pls kick github action again

ttwards commented:
Hi there, sorry for pausing so long, busy preparing for final exams,,, could you please review it again? Thanks a lot !!!!!

XenuIsWatching commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,16 +454,33 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
+	}
+	/* temp_int11 ranges in [-504, 496] */
+	/* the scale is 0.125°C/LSB = 125 micro degrees */
+	temp_micro = temp_int11 * 0.125 * 1000ULL + 23 * 1000000ULL;
```
this is using double percision floating point, can do it without double percision?

ttwards replied:
Should we eliminate the temp_micro intermediate variable and directly calculate val->val1 and val->val2 using these expressions:

/* temp_int11 ranges in [-504, 496] */
/* resolution: 125 micro degrees per count */
val->val1 = 23 + ((temp_int11 * 125) / 1000000);
val->val2 = (temp_int11 * 125) % 1000000;

Or just:
temp_micro = temp_int11 * 125 + 23 * 1000000;

ttwards replied:
Okay. But at the same time I wonder doesn't the compiler optimize it automatically? thx!

XenuIsWatching replied:
It wouldn't optimize something like that... but it looks good now thanks!
https://godbolt.org/z/E5WTvvxEK

ttwards replied:
I see! thanks a lot! I used to think that GCC would automatically determine variable types and treat all constant numbers as the same type.

XenuIsWatching reviewed (changes requested):
see comment

ubieda commented:
Thanks for your contribution!

XenuIsWatching commented:
@MaureenHelm Please review

avisconti commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,16 +454,33 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
-	if (ret < 0) {
+	if (!ret) {
+		temp_int11 = (temp_raw & 0xFF) << 3;
+	} else {
```
use LOG_ERR here?

avisconti commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,16 +454,33 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
-	/* the scale is 1/2^5/LSB = 31250 micro degrees */
-	temp_micro = BMI08X_TEMP_OFFSET * 1000000ULL + temp_raw * 31250ULL;
+	if (temp_raw == 0x80) {
+		/* temperature invalid */
```
use LOG_ERR?

avisconti commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,16 +454,33 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
+	ret = bmi08x_accel_word_read(dev, BMI08X_REG_TEMP_LSB, &temp_raw);
+	if (!ret) {
+		temp_int11 |= (temp_raw & 0xE0) >> 5;
+	} else {
```
LOG_ERR?

avisconti commented on drivers/sensor/bosch/bmi08x/bmi08x_accel.c:
```diff
@@ -454,16 +454,33 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
+	} else {
+		return ret;
+	}
+	if (temp_int11 > 1023) {
```
can you add a comment to clarify the meaning?

avisconti reviewed (commented):
LGTM, but just few suggestions to improve the code

avisconti commented:
Pls fix the CI

ttwards commented:
> Pls fix the CI

Sorry for that, clang-format in my local workspace doesn't meet the zephyr project requirements very well.

ubieda reviewed (approved):
Refreshing my +1
